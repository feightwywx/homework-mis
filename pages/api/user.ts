export type User = {
  isLoggedIn: boolean
  name: string
  userType: UserType
  token: string
}

export type UserType = 'student' | 'teacher'