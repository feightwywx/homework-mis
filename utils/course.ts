import { createMisConnection } from "./mysql";
import { Course } from "./types";

type CourseRow = {
  id: number;
  name: string;
  time: string;
  ended: number;
  teacherID: number;
  teacherName: string;
}

export async function getStudentCourses(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN course_student ON course.id = course_student.studentID 
    JOIN teacher ON course.teacherID = teacher.id
    WHERE course_student.studentID = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}

export async function getTeacherCourses(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN teacher ON course.teacherID = teacher.id
    WHERE teacher.id = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}

export async function getCourseByID(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN teacher ON course.teacherID = teacher.id
    WHERE course.id = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses[0];
}

export async function getCourses() {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.*
    FROM course`
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}
