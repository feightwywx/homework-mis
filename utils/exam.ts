import { ResultSetHeader } from "mysql2";
import { createMisConnection } from "./mysql";
import { Exam, ExamResult } from "./types";

type ExamRow = {
  id: number;
  name: string;
  time: string;
  endtime: string;
  location: string;
  courseID: number;
};

type ExamResultRow = {
  score: number | null;
};
type DetailedExamResultRow = ExamResultRow & {
  id: number;
  studentID: number;
  studentName: string;
};

export async function getExamsByCourseID(coid: number): Promise<Exam[]> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT *
    FROM exam
    WHERE courseID=?`,
    [coid]
  );
  await conn.end();

  const exams = rows as Array<ExamRow>;

  return exams;
}

export async function getExamByID(eid: number): Promise<Exam> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT *
    FROM exam
    WHERE id=?`,
    [eid]
  );
  await conn.end();

  const exams = (rows as Array<ExamRow>)[0];

  return exams;
}

export async function getStudentExamScore(
  uid: number,
  eid: number
): Promise<number | null> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT score
    FROM exam_result
    WHERE studentID=? AND examID=?;`,
    [uid, eid]
  );
  await conn.end();

  const score = (rows as Array<ExamResultRow>)[0].score;
  console.log("score", score);

  return score;
}

export async function getTeacherExamScore(eid: number): Promise<ExamResult[]> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT exam_result.id, studentID, student.name AS studentName, score
    FROM exam_result
    JOIN student ON student.id = exam_result.studentID
    WHERE exam_result.examID = ?;`,
    [eid]
  );
  await conn.end();

  const score = rows as Array<DetailedExamResultRow>;

  return score;
}

export async function updateExamScore(
  erid: number,
  score: number
): Promise<number> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `UPDATE exam_result SET score=? WHERE id=?;`,
    [score, erid]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows;
}

export async function insertExamAssign(courseID: number, title: string, location: string, time: string, endtime: string) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO exam (name, location, time, endtime, courseID) ' +
    'VALUES (?, ?, ?, ?, ?) ',
    [title, location, time, endtime, courseID]
  );
  await conn.end();

  return (rows as ResultSetHeader).insertId
}

export async function insertExamTarget(studentID: number, examID: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO exam_result (studentID, examID) ' +
    'VALUES (?, ?) ',
    [studentID, examID]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows
}
