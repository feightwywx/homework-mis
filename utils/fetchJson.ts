export default async function fetchJson<JSON = unknown>(
  input: RequestInfo, init?: RequestInit
): Promise<JSON> {
  const res = await fetch(input, init);
  const data = await res.json();

  if (res.ok) return data;

  throw new Error(res.statusText);
}