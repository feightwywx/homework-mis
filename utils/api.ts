import { NextApiRequest } from "next";
import { JsonResponse, ResponseResult, UserType } from "./types";
import { getId } from "./user";

export function successResponse<T = ResponseResult>(result?: T) {
  return {
    code: 0, 
    result: result ?? ''
  } as JsonResponse<T>
}

export function failResponse(code: number, result?: ResponseResult) {
  return {
    code, 
    result: result ?? ''
  } as JsonResponse
}

export async function parseIdFromReqest(req: NextApiRequest, userType: UserType) {
  let token = undefined;
  if (req.session.user?.token) {
    token = req.session.user?.token
  } else {
    token = await req.body.token;
  }

  return await getId(userType, token);
}

export const statusCode = {
  ROUTER_PARAM_REQUIRED: 100,
  ROUTER_PARAM_INVALID: 1001,
  BODY_PARAM_REQUIRED: 101,
  URL_PARAM_REQUIRED: 102,
  TOKEN_INVALID: 200,
  NUL_QUERY_DATA: 300,
  NUL_QUERY_PARAM: 301,
  BAD_UPDATE: 400
}
