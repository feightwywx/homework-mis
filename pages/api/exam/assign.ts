import { withIronSessionApiRoute } from "iron-session/next";
import { NextApiRequest, NextApiResponse } from "next";
import {
  failResponse,
  parseIdFromReqest,
  parseUserTypeFromRequest,
  statusCode,
  successResponse,
} from "../../../utils/api";
import { sessionOptions } from "../../../utils/session";
import {
  getExamByID,
  getStudentExamScore,
  getExamsByCourseID,
  getTeacherExamScore,
  insertExamAssign,
  insertExamTarget,
} from "../../../utils/exam";
import { ExamAssginRequestBody } from "../../../utils/types";
import { getCourseStudents } from "../../../utils/course";

export async function examScoreRoute(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const { title, location, time, endtime, courseID }: ExamAssginRequestBody =
    await req.body;

  if (!title) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, "title required"));
    return;
  }
  if (!location) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, "location required"));
    return;
  }
  if (!time) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, "time required"));
    return;
  }
  if (!endtime) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, "endtime required"));
    return;
  }
  if (!courseID) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, "courseID required"));
    return;
  }

  const id = await parseIdFromReqest(req);
  if (!id) {
    res.json(failResponse(statusCode.TOKEN_INVALID));
    return;
  }

  const userType = await parseUserTypeFromRequest(req);

  if (!userType || userType === "student") {
    res.json(failResponse(statusCode.USER_TYPE_NOT_SUPPORTED));
    return;
  }

  if (userType === "teacher") {
    const eid = await insertExamAssign(
      courseID,
      title,
      location,
      time,
      endtime
    );
    const target = await getCourseStudents(+courseID);

    if (!eid) {
      res.json(failResponse(statusCode.BAD_UPDATE));
      return;
    }

    let result = 0;
    await Promise.all(
      target.map(async (studentID) => {
        result += await insertExamTarget(studentID, eid);
      })
    );

    if (result) {
      res.json(successResponse(result));
    } else {
      res.json(failResponse(statusCode.NUL_QUERY_DATA));
    }
  }

  return;
}

export default withIronSessionApiRoute(examScoreRoute, sessionOptions);
