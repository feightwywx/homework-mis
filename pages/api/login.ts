import type { User } from './user';

import { withIronSessionApiRoute } from 'iron-session/next'
import { sessionOptions } from '../../utils/session'
import { NextApiRequest, NextApiResponse } from 'next'
import { getToken } from '../../utils/token';

export default async function loginRoute(req: NextApiRequest, res: NextApiResponse) {

  const { username, password, usertype } = await req.body

  res.json(await getToken(username, password, usertype));
}