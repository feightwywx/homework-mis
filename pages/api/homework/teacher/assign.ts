import { withIronSessionApiRoute } from "iron-session/next";
import { NextApiRequest, NextApiResponse } from "next";
import { failResponse, parseIdFromReqest, statusCode, successResponse } from "../../../../utils/api";
import { insertAssign, insertTarget } from "../../../../utils/homework";
import { sessionOptions } from "../../../../utils/session";

async function teacherAssignRoute(req: NextApiRequest, res: NextApiResponse) {
  const { title, assignment, deadline, target } = await req.body;

  if (!title) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, 'title required'));
    return;
  }
  if (!deadline) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, 'deadline required'));
    return;
  }
  if (!target) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, 'target required'));
    return;
  }

  const id = await parseIdFromReqest(req, 'teacher');
  if (!id) {
    res.json(failResponse(statusCode.TOKEN_INVALID));
    return;
  }

  const hwid = await insertAssign(+id, title, assignment, deadline);
  if (!hwid) {
    res.json(failResponse(statusCode.BAD_UPDATE))
    return;
  } else {
    let result = 0;
    (target as number[]).forEach(async value => {
      result += await insertTarget(hwid, value)
    })

    if (result) {
      res.json(successResponse({ affected: result }));
    } else {
      res.json(failResponse(statusCode.BAD_UPDATE));
    }
    return;
  }

}

export default withIronSessionApiRoute(teacherAssignRoute, sessionOptions)
