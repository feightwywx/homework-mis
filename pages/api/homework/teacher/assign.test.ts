import { createMocks } from 'node-mocks-http';
import type { NextApiRequest, NextApiResponse as OrigNextApiResp } from 'next';
import { teacherAssignRoute } from './assign';
import type { JsonResponse } from '../../../../utils/types';
import { initTestDatabase } from '../../../../test/utils/initTestDatabase';
import { statusCode } from '../../../../utils/api';

interface NextApiResponse<T = any> extends OrigNextApiResp<T> {
  _getJSONData: () => T;
}

describe('/api/homework/teacher/assign endpoint', () => {
  beforeAll(async () => {
    return initTestDatabase()
  })

  const token = '128f3a6727ead7a49aaf2caf4a7f66d29f4cf614ea36c2ce09ade39dfbd3e666af14eb906caf9f5aeddf04bdd52947e0';

  function mockRequestResponse(token?: string, title?: string, assignment?: string, deadline?: string, target?: number[]) {
    // @ts-expect-error
    const { req, res }: { req: NextApiRequest, res: NextApiResponse<JsonResponse> } = createMocks({
      method: 'GET',
      body: {
        token,
        title,
        assignment,
        deadline,
        target
      },
    })

    return { req, res }
  }

  it('if !title', async () => {
    const { req, res } = mockRequestResponse(token);
    await teacherAssignRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(statusCode.BODY_PARAM_REQUIRED);
    expect(res._getJSONData().result).toEqual('title required');
  })

  it('if !title', async () => {
    const { req, res } = mockRequestResponse(token, 'title');
    await teacherAssignRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(statusCode.BODY_PARAM_REQUIRED);
    expect(res._getJSONData().result).toEqual('assignment required');
  })

  it('if !deadline', async () => {
    const { req, res } = mockRequestResponse(token, 'title', 'assignment');
    await teacherAssignRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(statusCode.BODY_PARAM_REQUIRED);
    expect(res._getJSONData().result).toEqual('deadline required');
  })

  it('if !target', async () => {
    const { req, res } = mockRequestResponse(token, 'title', 'assignment', 'deadline');
    await teacherAssignRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(statusCode.BODY_PARAM_REQUIRED);
    expect(res._getJSONData().result).toEqual('target required');
  })

  it('returns TOKEN_INVALID if !token', async () => {
    const { req, res } = mockRequestResponse(undefined, 'title', 'assignment', 'deadline', []);
    await teacherAssignRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(statusCode.TOKEN_INVALID);
  })

  it('if ok', async () => {
    const { req, res } = mockRequestResponse(token, 'title', 'assignment', '2023-10-25 22:50:02', [1]);
    await teacherAssignRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData()).toMatchSnapshot();
  })
})