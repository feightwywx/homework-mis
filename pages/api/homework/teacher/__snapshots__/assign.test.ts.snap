// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/homework/teacher/assign endpoint if ok 1`] = `
{
  "code": 0,
  "result": {
    "affected": 1,
  },
}
`;
